### PR TITLE
test: Rework ci (Use travis only as fallback env)

### PR DESCRIPTION
**Pull request by maflcko**
This moves the `.travis` folder to `ci` and removes dependence on travis, so that the test script can be run anywhere.

### CODE REVIEW

**Comment by practicalswift** [comment]
@MarcoFalke Could you provide some additional context? With this PR merged tests would still run in both Appveyor and Travis, right? What is the plan? :-)

**Comment by MarcoFalke** [comment]
The plan is to be flexible, to be able to switch away from travis any time without much hassle. Also, this makes it easier to run the ci locally.

File: ci/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+## ci scripts
+
+This directory contains scripts for each build step in each build stage.
+
+Currently three stages `lint`, `extended_lint` and `test` are defined. Each stage has it's own lifecycle, similar to the
```
Comment (sipa) [inline ci/README.md]: its

**Comment by promag** [comment]
Concept ACK.

**Comment by MarcoFalke** [comment]
Fixed typo in doc and added a note on how to run the ci locally

**Review by fanquake — COMMENTED** [review]
Concept ACK. 

Have tested running the CI locally on macOS using Docker and `./ci/test_run_all.sh`. Everything seems to work as expected.

It's a little annoying that we have to bring more bash scripts into this repo, but I guess they can just sit, and we wont have to modify/touch them at all.

edit: Could also add `out/` and `qa-assets/` to `.gitignore`.

File: ci/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+[Travis CI lifecycle](https://docs.travis-ci.com/user/job-lifecycle#the-job-lifecycle).  Every script in here is named
+and numbered according to which stage and lifecycle step it belongs to.
+
+### Running a stage locally
+
+To allow for a wide range of tested environments, but also ensure reproducibility to some extent, the test stage
+requires `docker` to be installed. To install all requirements on Ubuntu, run
+
+```
+sudo apt install docker.io ccache bash git
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply ci/README.md]: Huh? According to https://packages.ubuntu.com/xenial/docker.io it is the latest version. We don't need any fancy new features (other than `--mount`), so it should be fine as is.

[the remark this answers is not in the document]
Comment (fanquake) [reply ci/README.md]: Interesting. From the [Docker documentation](https://docs.docker.com/install/linux/docker-ce/ubuntu/):
> Older versions of Docker were called docker, docker.io , or docker-engine. If these are installed, uninstall them:
> ```bash
> sudo apt-get remove docker docker-engine docker.io containerd runc
> ```

I guess I don't mind either way, just wasn't sure after looking at those docs.

**Comment by MarcoFalke** [comment]
> Have tested running the CI locally on macOS using Docker and ./ci/test_run_all.sh. Everything seems to work as expected.

Thanks for testing on mac!

So this is confirmed to be working on:
* Locally, macOS host (fanquake)
* travis, xenial host (run as part of this pull)
* cirrus ci, bionic host (https://cirrus-ci.com/task/5077011800784896)
* GitHub ci, bionic host (https://github.com/MarcoFalke/bitcoin/runs/190746548)
* Locally, bionic host (MarcoFalke)

**Comment by instagibbs** [comment]
> Also, this makes it easier to run the ci locally.

Concept ACK!

File: ci/retry/retry [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    #TODO: check if user would rather use one of the other possible dependencies: python, ruby, bc, dc
+    sleep_time=`awk "BEGIN {t = $min_sleep * $(( (1<<($attempts -1)) )); print (t > $max_sleep ? $max_sleep : t)}"`
+  fi
+}
+
+__log_out() {
+  echo "$1" 1>&2
+}
+
+# Paramters: max_tries min_sleep max_sleep constant_sleep fail_script EXECUTION_COMMAND
```
Comment (practicalswift) [inline ci/retry/retry]: Should be "parameters" :-)

File: ci/retry/retry [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  help()
+  {
+    local retry=$(basename $0)
+    cat <<EOF
+Usage: $retry [options] -- execute command
+    -h, -?, --help
+    -v, --verbose                    Verbose output
+    -t, --tries=#                    Set max retries: Default 10
+    -s, --sleep=secs                 Constant sleep amount (seconds)
+    -m, --min=secs                   Exponenetial Backoff: minimum sleep amount (seconds): Default 0.3
```
Comment (practicalswift) [inline ci/retry/retry]: Should be exponential -- applies throughout PR :-)

File: .cirrus.yml [hunk 5]
```diff
@@ -27,3 +27,26 @@ task:
     - gmake check ${MAKEJOBS} VERBOSE=1
   functional_test_script:
     - ./test/functional/test_runner.py --jobs 9 --ci --extended --exclude feature_dbcrash --combinedlogslen=1000 --quiet --failfast
+task:
```
Comment (practicalswift) [inline .cirrus.yml]: Duplication of key `task` in mapping?

File: .cirrus.yml [hunk 6]
```diff
@@ -27,3 +27,26 @@ task:
...
+  env:
+    MAKEJOBS: "-j9"
+    RUN_CI_ON_HOST: "1"
+    CCACHE_SIZE: "200M"
+    CCACHE_DIR: "/tmp/ccache_dir"
+  ccache_cache:
+    folder: "/tmp/ccache_dir"
+  depends_cache:
+    folder: "/tmp/cirrus-ci-build/depends/built"
+    folder: "/tmp/cirrus-ci-build/depends/sdk-sources"
```
Comment (practicalswift) [inline .cirrus.yml]: Duplication of key `folder` in mapping?

File: ci/retry/retry (hunk 3 above)
Comment (fanquake) [reply ci/retry/retry]: This is an [upstream script](https://github.com/kadwanev/retry), so we don't need to fix it's issues here. Regardless, I've [PR'd](https://github.com/kadwanev/retry/pull/15) them upstream.

File: .cirrus.yml (hunk 5 above)
Comment (MarcoFalke) [reply .cirrus.yml]: Yes, this is the correct syntax:

https://cirrus-ci.org/guide/writing-tasks/#task-execution-dependencies

**Comment by MarcoFalke** [comment]
Rebased with three-way-merge due to change in tangent line

**Comment by fanquake** [comment]
I'm now seeing an error running this:
```
configure.ac:40: installing 'build-aux/missing'
src/Makefile.am: installing 'build-aux/depcomp'
travis_fold:end:autogen

travis_fold:start:configure
configure: loading site script /Users/michael/github/bitcoin/depends/x86_64-unknown-linux-gnu/share/config.site
configure: loading cache config.cache
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for x86_64-unknown-linux-gnu-strip... no
checking for strip... strip
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
configure: error: source directory already configured; run "make distclean" there first
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by Bitcoin Core configure 0.18.99, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ../configure --cache-file=config.cache --disable-dependency-tracking --prefix=/Users/michael/github/bitcoin/depends/x86_64-unknown-linux-gnu --bindir=/Users/michael/github/bitcoin/out/x86_64-unknown-linux-gnu/bin --libdir=/Users/michael/github/bitcoin/out/x86_64-unknown-linux-gnu/lib
```

**Comment by MarcoFalke** [comment]
It says `source directory already configured; run "make distclean" there first`. Have you tried that?

**Comment by MarcoFalke** [comment]
Specifically `cd /Users/michael/github/bitcoin && make distclean`.

The ci is generally assuming the the folder is a fresh git clone. If needed, that can be changed later on, but better not part of this pull.

**Comment by fanquake** [comment]
> It says source directory already configured; run "make distclean" there first. Have you tried that?

Shouldn't that all be handled by the script? If not, then some docs about why you can't run the tests twice in a row would be handy.

**Comment by MarcoFalke** [comment]
Done

**Review by fanquake — APPROVED** [review]
Thanks ACK fa9404d76d16c618dbd2c19973246900918f9e6c - Maybe @sipsorcery or @NicolasDorier might also want to test here, then we'd have all OS's covered?

**Comment by MarcoFalke** [comment]
I don't think docker on windows can run Ubuntu containers

**Comment by MarcoFalke** [comment]
Rebased without any changes due to overly sensitive merge conflict detection on GitHub. git range-diff should be empty: `git range-diff bitcoin/master fa9404d76d faeacf3269`

**Comment by sipsorcery** [comment]
> I don't think docker on windows can run Ubuntu containers

It can since https://docs.docker.com/docker-for-windows/wsl-tech-preview/

````
aaron@pcdodo:~$ uname -a
Linux pcdodo 4.19.57-microsoft-standard #1 SMP Wed Jul 3 20:47:25 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
aaron@pcdodo:~$ docker run ubuntu echo "I  love Windows"; lsb_release -a
I  love Windows
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.2 LTS
Release:        18.04
Codename:       bionic
````
Actually it was already possible to use ubuntu docker containers on Windows but now it  possible directly on top of the Windows Subsystem for Linux kernel.

**Comment by sipsorcery** [comment]
> Thanks ACK [fa9404d](https://github.com/bitcoin/bitcoin/commit/fa9404d76d16c618dbd2c19973246900918f9e6c) - Maybe @sipsorcery or @NicolasDorier might also want to test here, then we'd have all OS's covered?

@fanquake I looked over the PR and successfully ran my appveyor build with the PR merged. Doesn't seem to be anything Windows related in this PR? Ony travis and shell scripts. Or did you testing the script changes with the WSL/Mingw builds on Windows?

**Comment by fanquake** [comment]
Thanks @sipsorcery. I also thought it was possible to run on Windows.

> Or did you testing the script changes with the WSL/Mingw builds on Windows?

Yea I just meant checking out this PR, and testing/running the scripts, to make sure they were working in your Windows development setups.

**Pull request merged**